### PR TITLE
ARROW-8379: [R] Investigate/fix thread safety issues (esp. Windows)

### PR DESCRIPTION
**Pull request by nealrichardson**
One side effect of this change is that it actually re-enables lots of dataset tests on Windows, just not for 32-bit rtools35

### CODE REVIEW

File: r/tests/testthat/helper-arrow.R [hunk 1]
```diff
@@ -18,12 +18,6 @@
 # Wrap testthat::test_that with a check for the C++ library
 options(..skip.tests = !arrow:::arrow_available())
 
-if (tolower(Sys.info()[["sysname"]]) == "windows") {
-  # For now, disable multithreading by default on Windows
-  # See https://issues.apache.org/jira/browse/ARROW-8379
-  options(arrow.use_threads = FALSE)
-}
```
Comment (jonkeane) [inline r/tests/testthat/helper-arrow.R:25]: I've copy/pasted this chunk into ticket https://issues.apache.org/jira/browse/ARROW-14159?focusedCommentId=17422217&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17422217 so it's obvious (one way to) make this happen when we get there

**Review by jonkeane — COMMENTED** [review]
@github-actions crossbow submit -g r

**Comment by jonkeane** [comment]
@github-actions crossbow submit -g r

lol apparently the PR review kind of comment doesn't count?

**Comment by nealrichardson** [comment]
I'm going to merge just so we get more testing on this before the release (i.e. if there are flaky problems with Datasets on Windows other than 32-bit rtools 35, which we haven't been testing). Happy to revisit/tweak this if anyone has concerns after merging.

**Pull request closed without merging**